resolver/resolver.cc

```diff
@@ -1120,6 +1155,185 @@ class ResolveTypeMembersWalk {
+                return;
+            } else {
+                // We do some normalization here to ensure that the file / line we report the error on doesn't
+                // depend on the order that we traverse files nor the order we traverse within a file.
+                auto priorLoc = prior.data(ctx)->loc();
+                core::Loc reportOn;
+                core::Loc errorLine;
+                core::Loc thisLoc = core::Loc(job.file, uid->loc);
+                if (thisLoc.file() == priorLoc.file()) {
+                    reportOn = thisLoc.beginPos() < priorLoc.beginPos() ? thisLoc : priorLoc;
+                    errorLine = thisLoc.beginPos() < priorLoc.beginPos() ? priorLoc : thisLoc;
+                } else {
+                    reportOn = thisLoc.file() < priorLoc.file() ? thisLoc : priorLoc;
+                    errorLine = thisLoc.file() < priorLoc.file() ? priorLoc : thisLoc;
+                }
+
+                if (auto e = ctx.state.beginError(reportOn, core::errors::Resolver::DuplicateVariableDeclaration)) {
+                    e.setHeader("Redeclaring variable `{}` with mismatching type", uid->name.show(ctx));
+                    e.addErrorLine(errorLine, "Previous declaration is here:");
+                }
+                return;
+            }
+        }
+        core::SymbolRef var;
+
+        if (uid->kind == ast::UnresolvedIdent::Kind::Class) {
+            var = ctx.state.enterStaticFieldSymbol(core::Loc(job.file, uid->loc), scope, uid->name);
+        } else {
+            var = ctx.state.enterFieldSymbol(core::Loc(job.file, uid->loc), scope, uid->name);
+        }
+
+        var.data(ctx)->resultType = cast->type;
+        return;
+    }
+
+    // Resolve the type of the rhs of a constant declaration. This logic is
+    // extremely simplistic; We only handle simple literals, and explicit casts.
+    //
+    // We don't handle array or hash literals, because intuiting the element
+    // type (once we have generics) will be nontrivial.
+    static core::TypePtr resolveConstantType(core::Context ctx, const ast::TreePtr &expr) {
+        core::TypePtr result;
+        typecase(
+            expr, [&](const ast::Literal &a) { result = a.value; },
+            [&](const ast::Cast &cast) {
+                if (cast.type != nullptr && cast.cast != core::Names::let() &&
+                    cast.cast != core::Names::uncheckedLet()) {
+                    if (auto e = ctx.beginError(cast.loc, core::errors::Resolver::ConstantAssertType)) {
+                        e.setHeader("Use `{}` to specify the type of constants", "T.let");
+                    }
+                }
+                result = cast.type;
+            },
+            [&](const ast::InsSeq &outer) { result = resolveConstantType(ctx, outer.expr); },
+            [&](const ast::TreePtr &expr) {});
+        return result;
+    }
+
+    [[nodiscard]] static core::TypePtr tryResolveStaticField(core::Context ctx, ResolveStaticFieldItem &job) {
```

Comment:
Why did you only mark this one as `[[nodiscard]]`?

Reply:
Good point. I marked the others as well.